### PR TITLE
docgen: Replace doctrine with comment-parser

### LINKED ISSUE #18045: docgen: remove doctrine
`doctrine` is the JSDoc parser we use in `docgen` to get structured information out of the JSDoc comments. It was created and maintained by the ESLint people, although it's [no longer actively developed](https://eslint.org/blog/2018/11/jsdoc-end-of-life). We should migrate out of `doctrine` for that reason.

The suggested replacement ([eslint-plugin-jsoc](https://github.com/gajus/eslint-plugin-jsdoc)) uses a different [JSDoc parser](https://github.com/jsdoctypeparser/jsdoctypeparser) with support for many other syntaxes (typescript, etc). We should look into trying `jsdoctypeparser` or any other that's actively developed.

### LINKED ISSUE #26971: Docgen: Some types omitted
Some types may be omitted from docgen output. For example:

```js
/** @type {Record<string,true|undefined>} */
```

https://github.com/WordPress/gutenberg/blob/0545e7d009316338e14e11e53e865723165ce6b8/packages/deprecated/src/index.js#L6-L12

Was rendered as:

```md
_Type_

-   `Record<string,(|undefined)>` 
```

_Originally posted by @sirreal in https://github.com/WordPress/gutenberg/pull/26429#r515412062_

### PR DESCRIPTION
**Pull request by sarayourfriend**
## Description
Replaces the usage of `doctrine` with `comment-parser`.

Notable differences:

1. Tuples are no longer surrounded by parens, they switch from `(string|Object)` to `string|Object`.
2. Arrays switch from `Array<Object>` to `Object[]`.
3. No types are normalized, they're left the way they are in the declaration rather than being parsed and then normalized. Examples of some changes related to this: nullable types can be either `?type` to `type?` and that optional types can be either `[type]` or `type=`, depending on how the author declared them in the JSDoc source comment.

Notable issues with `comment-parser`.
1. The library parses the following:
```js
/**
 * @return {string} The description for a return.
 */
```
as
```json
{
        "tag": "return",
        "name": "The",
        "type": "string",
        "optional": false,
        "description": "description for a return.",
}
```
Note that the description of the return is split into two... this appears to be a difference in convention, some folks name their return values. We tend _not_ to do so. Regardless, the library does this for all tags so you'll see something resembling `` `${tag.name} ${tag.description}` `` places where previously we could just use the description. I don't personally see this as an issue, it's just a quirk of the library, but I could see this getting in the way of adopting this approach if folks think it's too fragile or something like that.

Notable is that I don't think this is a bug in the library, I think it's just the way it parses based on some conventions. It's a little silly for tags like `@see` but I think it's livable.

## How has this been tested?
Tests pass and documentation generation still works as expected (with some minor changes given the new parser)

## Types of changes
Non-breaking changes

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. 

Fixes #26971
Fixes #18045

### CODE REVIEW

File: docs/designers-developers/developers/data/data-core-keyboard-shortcuts.md [hunk 1]
```diff
@@ -79,11 +79,11 @@ _Parameters_
 
 -   _state_ `Object`: Global state.
 -   _name_ `string`: Shortcut name.
--   _representation_ (unknown type): Type of representation (display, raw, ariaLabel).
+-   _representation_ `keyof FORMATTING_METHODS`: Type of representation (display, raw, ariaLabel).
```
Comment (gziolo) [inline docs/designers-developers/developers/data/data-core-keyboard-shortcuts.md:82]: Nice 👍🏻

File: packages/data-controls/README.md [hunk 2]
```diff
@@ -74,7 +74,7 @@ resolvers,
 
 _Parameters_
 
--   _paths_ (unknown type): 
+-   _paths_ ``:
```
Comment (gziolo) [inline packages/data-controls/README.md:77]: It still isn't working.

File: packages/date/README.md [hunk 3]
```diff
@@ -22,14 +22,13 @@ Formats a date (like `date()` in PHP).
 
 _Related_
 
--   <https://en.wikipedia.org/wiki/List_of_tz_database_time_zones>
--   <https://en.wikipedia.org/wiki/ISO_8601#Time_offsets_from_UTC>
+-   <https://en.wikipedia.org/wiki/List_of_tz_database_time_zones-> <https://en.wikipedia.org/wiki/ISO_8601#Time_offsets_from_UTC>
```
Comment (gziolo) [inline packages/date/README.md:25]: Can we split it into multiple lines again?

File: packages/deprecated/README.md [hunk 4]
```diff
@@ -68,7 +68,7 @@ message is only logged once.
 
 _Type_
 
--   `Record<string,(|undefined)>` 
+-   `Record<string,true|undefined>`
```
Comment (gziolo) [inline packages/deprecated/README.md:71]: Nice 👍🏻

File: packages/edit-post/README.md [hunk 5]
```diff
@@ -48,6 +48,7 @@ Renders a new item in the block settings menu.
 
 _Usage_
 
+<caption>ES5</caption>
```
Comment (gziolo) [inline packages/edit-post/README.md:51]: I think it should strip this line with the `<caption>`. If I recall properly it has some special handling in the Block Editor Handbook already.

File: packages/edit-post/README.md [hunk 6]
```diff
@@ -63,30 +64,29 @@ function MyPluginBlockSettingsMenuItem() {
 		{
 			allowedBlocks: [ 'core/paragraph' ],
 			icon: 'dashicon-name',
-			label: __( 'Menu item text' ),
+			label: \_\_( 'Menu item text' ),
```
Comment (gziolo) [inline packages/edit-post/README.md:67]: It's wrongly processed.

File: packages/edit-post/README.md [hunk 7]
```diff
@@ -235,29 +236,28 @@ function MyPluginPostPublishPanel() {
 		PluginPostPublishPanel,
 		{
 			className: 'my-plugin-post-publish-panel',
-			title: __( 'My panel title' ),
+			title: **( 'My panel title' ),
 			initialOpen: true,
 		},
-		__( 'My panel content' )
+		**( 'My panel content' )
```
Comment (gziolo) [inline packages/edit-post/README.md:242]: It's wrongly processed as well.

File: packages/edit-post/README.md [hunk 8]
```diff
@@ -181,30 +182,29 @@ function MyButtonMoreMenuItem() {
...
 
-const MyButtonMoreMenuItem = () => (
-	<PluginMoreMenuItem
-		icon={ more }
-		onClick={ onButtonClick }
-	>
-		{ __( 'My button title' ) }
-	</PluginMoreMenuItem>
-);
-```
```
Comment (gziolo) [inline packages/edit-post/README.md:207]: It removes closing "```"

File: packages/element/README.md [hunk 9]
```diff
@@ -180,7 +180,7 @@ _Related_
 
 _Parameters_
 
--   _child_ (unknown type): Any renderable child, such as an element, string, or fragment.
+-   _child_ `import('./react').WPElement`: Any renderable child, such as an element, string, or fragment.
```
Comment (gziolo) [inline packages/element/README.md:183]: Nice!

File: packages/element/README.md [hunk 10]
```diff
@@ -265,9 +265,7 @@ This is the same concept as the React Native implementation.
 
 _Related_
 
--   <https://facebook.github.io/react-native/docs/platform-specific-code#platform-module>
-
-Here is an example of how to use the select method:
+-   <https://facebook.github.io/react-native/docs/platform-specific-code#platform-module> Here is an example of how to use the select method:
```
Comment (gziolo) [inline packages/element/README.md:268]: Line breaks get removed.

File: packages/keycodes/README.md [hunk 11]
```diff
@@ -134,7 +134,7 @@ depending on platform.
 
 _Type_
 
--   (unknown type) 
+-   `WPModifierHandler< ( isApple: () => boolean ) => WPModifierPart[]>`
```
Comment (gziolo) [inline packages/keycodes/README.md:137]: ❤️

**Review by gziolo — COMMENTED** [review]
Awesome work. It looks very promising. There are still some issues to resolve but overall it'd be a great improvement as we start using more TypeScript syntax for types.

I left some comments where I could spot something that concerns me and places where you can see TS types properly included.

I would love to hear also from @nosolosw how does he feel about the changes proposed since he authored `@wordpress/docgen` and he has the most experience here.

**Comment by sirreal** [comment]
I've only scanned this but it looks really promising. I wanted to share [TSDoc](https://tsdoc.org/) which is worth testing and exploring. It may either be usable directly or a good source of inspiration. I think this line is important and represents some common frustrations with JSDoc:

> **_Why can’t JSDoc be the standard?_** The JSDoc grammar is not rigorously specified, but rather inferred from the behavior of a particular implementation. The majority of the standard JSDoc tags are preoccupied with providing type annotations for plain JavaScript, which is not a primary concern for a strongly-typed language such as TypeScript.

**Comment by saramarcondes** [comment]
> I wanted to share TSDoc which is worth testing and exploring.

Ooooo, I originally started to re-write docgen using `typescript` as the parser instead of `babel` because I didn't realize there was a way to just parse the doc comments with the TS parser, but the `TSDocParser` class looks perfect for that. I'll do some tests and see if I can make it work. It may help resolve some of the strange issues that `comment-parser` has too.

File: packages/data-controls/README.md (hunk 2 above)
Comment (saramarcondes) [reply packages/data-controls/README.md:77]: It is missing the type: https://github.com/WordPress/gutenberg/blob/d9e7541fe29c0022f2f5de2570ec60bee670aede/packages/data-controls/src/index.js#L106

File: packages/date/README.md (hunk 3 above)
Comment (saramarcondes) [reply packages/date/README.md:25]: Whoops, yes! I found the bug I introduced that caused this.

File: packages/edit-post/README.md (hunk 5 above)
Comment (saramarcondes) [reply packages/edit-post/README.md:51]: Where did this happen before?

File: packages/edit-post/README.md (hunk 8 above)
Comment (saramarcondes) [reply packages/edit-post/README.md:207]: I'm really confused why this is happening. The description has the trailing ```` ``` ```` in it when I inspect it, I have no idea why it's being removed.

To make matters worse, when I arbitrarily append a random string like `asdf` to the end of the description, the ```` ``` ```` suddenly appears with `asdf` appended. It's very odd.

File: packages/edit-post/README.md (hunk 5 above)
Comment (gziolo) [reply packages/edit-post/README.md:51]: @nosolosw should have an answer 😃

File: packages/docgen/src/test/formatter-markdown.js [hunk 12]
```diff
@@ -38,36 +39,4 @@ describe( 'Formatter', () => {
 			'# API docs\n\n<a name="myDeclaration" href="#myDeclaration">#</a> **myDeclaration**\n\nMy declaration example.\n\n*Parameters*\n\n- *firstParam* `number`: First declaration parameter.\n\n*Returns*\n\n- `number`: The result of the declaration.\n'
 		);
 	} );
-
-	it( 'handles unknown types from parse error', () => {
```
Comment (nosolosw) [inline packages/docgen/src/test/formatter-markdown.js:42]: What happens when there's a parser error using the new setup?

Comment (nosolosw) [reply packages/docgen/src/test/formatter-markdown.js:42]: I saw there's a related test removed in `get-intermediate-representation` as well.

File: packages/edit-post/README.md (hunk 5 above)
Comment (nosolosw) [reply packages/edit-post/README.md:51]: Did some digging on the `<captions>`. IIRC, the intention was to use them to present tabbed code, like in [this example](https://developer.wordpress.org/block-editor/tutorials/block-tutorial/writing-your-first-block-type/#registering-the-block) for ES5/ESNext. I don't know if we've ever implemented that, though. This is the [issue](https://github.com/WordPress/gutenberg/issues/14331).  Relevant PRs: https://github.com/WordPress/gutenberg/pull/14271 https://github.com/WordPress/gutenberg/pull/14263 https://github.com/WordPress/gutenberg/pull/21405

Additionally, we have to make sure that the comments look good in different places:

- block editor: https://developer.wordpress.org/block-editor/packages/packages-edit-post/#PluginBlockSettingsMenuItem
- npm https://www.npmjs.com/package/@wordpress/edit-post#PluginBlockSettingsMenuItem
- github https://github.com/WordPress/gutenberg/tree/master/packages/edit-post#PluginBlockSettingsMenuItem

I'd think that it's fine to remove the `<caption>` from the comments if they are problematic for `comment-parser`, as I don't see them in use anywhere.

**Comment by nosolosw** [comment]
I've updated the issue description with two other cases in which the docs changed (nullable and optional types).

File: packages/blocks/README.md [hunk 13]
```diff
@@ -379,15 +379,15 @@ _Parameters_
 
 _Returns_
 
--   `(Array<WPBlockVariation>|void)`: Block variations.
+-   `(WPBlockVariation[]|void)`: Block variations.
```
Comment (nosolosw) [inline packages/blocks/README.md:382]: In this case, the tuples still get the parens :thinking:

**Comment by nosolosw** [comment]
This is promising! Thanks for taking up on replacing `doctrine`. I've done a first pass at this and left some comments in addition to what Greg already mentioned. Can do a second pass when you're ready.

---

> I began exploring this avenue but it would require a significant refactor re-write of most of docgen to work. It didn't seem worth it when there was a much simpler interim solution (though I would love to dive into the work of re-writing docgen with TypeScript, I unfortunately do not have that kind of time on my hands at the moment).

Not for this PR, but I have some thoughts on the topic of upgrading / creating a replacement for `docgen` that is based on TSDoc instead of JSDoc. I thought it'd be useful to share, in case you have time to work on this in the future:

- We've experienced problems in the past when using different parsers for documentation (`espree`) and code (`babel`), hence why we switched to use the same everywhere. It was almost a [one-liner](https://github.com/WordPress/gutenberg/commit/6c26cdf1f97307b478e37d731351c4b1125e902b#diff-f8a16bfa70ef288f1f17d3d93f8888aae22fb5728670fcf0e8cd62625e260f90) change. If we were to split paths again, we should be aware of potential issues (one parser introduces support for some things unsupported by the other) and making sure the code is organized in a way that merging paths is easy, should we need it.

- The current approach is based on parsing individual files. Other approaches, like the jsdoc tool, parse the whole codebase instead. I believe the typescript parser does the same? Parsing individual files fails to integrate types that are defined in a different file, see https://github.com/WordPress/gutenberg/issues/15186 I think solving this long-standing issue could be one of the best use cases to address if we're thinking about rewrites.

File: packages/docgen/src/test/formatter-markdown.js (hunk 12 above)
Comment (saramarcondes) [reply packages/docgen/src/test/formatter-markdown.js:42]: I'll do some digging into the library to check and see how to trigger an error. This new library is pretty "simple" in that it doesn't seem to have many expectations about how the input is formatted except for the additional `/*` and `/` that needed to be prepended/appended to the comment input string.

**Comment by saramarcondes** [comment]
> I've updated the issue description with other cases in which the docs changed (nullable and optional types, arrays).

FWIW, these changes are due to differences in how the types are documented, all the library does is copy the type out of the curly braces (unlike the previous library which attempted to normalize them).

If we want to normalize these types, we could add that too but it would take significant work. There's an issue (#22062) to make these decisions that has almost no activity.

Personally I think it's a lost cause to try to normalize them during parse time, we might be better off spending that time normalizing the way they're actually recorded in the annotations themselves.

File: packages/blocks/README.md (hunk 13 above)
Comment (saramarcondes) [reply packages/blocks/README.md:382]: This is another example of the parser just copying the way the type is annotated out of the curly braces: https://github.com/WordPress/gutenberg/blob/1f5af853b83999f773d1e10ed250f06a294c73df/packages/blocks/src/api/registration.js#L537

**Comment by saramarcondes** [comment]
@nosolosw and @gziolo, do y'all have time to give this another peek sometime soon? I think it's ready for another review (all the feedback has been addressed).

File: packages/data-controls/README.md (hunk 2 above)
Comment (nosolosw) [reply packages/data-controls/README.md:77]: Can we remove the `@param paths` from here? I'm not familiar with this part of the code but the param doesn't make sense as it is.

File: packages/docgen/src/test/get-intermediate-representation.js [hunk 14]
```diff
@@ -50,37 +50,6 @@ describe( 'Intermediate Representation', () => {
 		} );
 	} );
 
-	it( 'type parse error handling', () => {
```
Comment (nosolosw) [inline packages/docgen/src/test/get-intermediate-representation.js:53]: I guess this makes sense to remove: as I understand it, `comment-parser` does actually not parse the type, just passes it through.

File: packages/docgen/src/test/formatter-markdown.js (hunk 12 above)
Comment (nosolosw) [reply packages/docgen/src/test/formatter-markdown.js:42]: Same as below: I guess it makes sense to remove this. As I understand it, comment-parser does actually not parse the type, just passes it through.

File: packages/docgen/src/test/get-intermediate-representation.js (hunk 14 above)
Comment (saramarcondes) [reply packages/docgen/src/test/get-intermediate-representation.js:53]: I did add a new error case just now for when the parsing fails to parse it as a JSDoc comment (in the case of malformed comment, for example). I don't know if it's possible IRL but was easy to test for and caught an null pointer exception that was possible in that case.

File: packages/element/README.md (hunk 10 above)
Comment (nosolosw) [reply packages/element/README.md:268]: The example intro is bundled as part of the `@see` tag description while is a separate thing. I presume `comment-parser` assigns all the text within tags to the previous tag, hence the error.

One thing we can do to fix this is moving the intro text `Here is an example of how to use the select method:` below the `@example` tag. That way it works fine.

File: packages/escape-html/README.md [hunk 15]
```diff
@@ -43,18 +43,7 @@ Returns an escaped attribute value.
...
-"[...] the text cannot contain an ambiguous ampersand [...] must not contain
-any literal U+0022 QUOTATION MARK characters (")"
-
-Note we also escape the greater than symbol, as this is used by wptexturize to
-split HTML strings. This is a WordPress specific fix
-
-Note that if a resolution for Trac#45387 comes to fruition, it is no longer
-necessary for `__unstableEscapeGreaterThan` to be used.
-
-See: <https://core.trac.wordpress.org/ticket/45387>
```
Comment (nosolosw) [inline packages/escape-html/README.md:57]: In the source, we can change `See: https://core.trac.wordpress.org/ticket/45387` by `@see https://core.trac.wordpress.org/ticket/45387` so the link to track is listed in a separate line.

Any ideas about how to fix the formatting of the comment @saramarcondes ? Not a blocker, but it's definitely not very nice.

**Review by oandregal — APPROVED** [review]
Thanks for working on this @saramarcondes! This is ready in my view.

There are a few weird cases in which the spacing is removed or the text added to the previous tag. Added some suggestions to fix this, it can be done here or in follow-ups.

We also need to update the changelog for the package release. Sharing my thoughts on this:

This PR replaces `doctrine`, a deprecated library, with `comment-parser`. The major change to our public API is that docgen no longer parses and structures the types (because `comment-parser` doesn't do it), it passes through whatever the author has written:

- This flexibility works great for types that aren't part of the [JSDoc spec](jsdoc.app/). For example, adding some typescript `import` syntax.

- It has the drawback of inconsistency: we are already seeing the effects of this with nullable or optional types, that different people write differently. Those differences are going to increase over time: I wonder if we can do anything to converge syntaxes (lint rule, etc?).

Given this context, it seems to me we should do a major release for `docgen`. @gziolo @saramarcondes thoughts?

File: packages/escape-html/README.md (hunk 15 above)
Comment (saramarcondes) [reply packages/escape-html/README.md:57]: It's interesting. This new parser has a "preserve" spacing argument but it only seems to respect that _sometimes_. I'll take a look at this, you're right that this doesn't look nice. Thanks for catching it.

Comment (saramarcondes) [reply packages/escape-html/README.md]: Found a fix for it, `cleanSpaces` also strips out new lines so we just shouldn't pass it through there.

**Comment by gziolo** [comment]
> We also need to update the changelog for the package release.


It definitely should be a major version bump. It's a different parser now and it will produce different output in most of the cases 👍🏻 

There is one failing unit test that needs to be fixed to unblock merging:

```
Summary of all failing tests
FAIL packages/docgen/src/test/get-jsdoc-from-token.js
  ● JSDoc › extracts description and tags

    expect(received).toMatchObject(expected)

    - Expected  - 3
    + Received  + 7

    @@ -25,20 +25,24 @@
            "optional": false,
            "tag": "param",
            "type": "number",
          },
          Object {
    -       "description": "The second param to add.",
    +       "description": "The second param to add.
    + ",
            "name": "secondParam",
            "tag": "param",
            "type": "number",
          },
          Object {
    -       "description": "```js
    +       "description": " 
    +
    + ```js
      const addResult = sum( 1, 3 );
      console.log( addResult ); // will yield 4
    - ```",
    + ```
    + ",
            "tag": "example",
          },
          Object {
            "description": "result of adding the two params.",
            "name": "The",

      15 | 				],
      16 | 			} )
    > 17 | 		).toMatchObject( {
         | 		  ^
      18 | 			description: 'A function that adds two parameters.\n',
      19 | 			tags: [
      20 | 				{

      at Object.<anonymous> (packages/docgen/src/test/get-jsdoc-from-token.js:17:5)

File: packages/edit-post/src/components/block-settings-menu/plugin-block-settings-menu-item.js [hunk 16]
```diff
@@ -40,7 +40,6 @@ const shouldRenderItem = ( selectedBlocks, allowedBlocks ) =>
  * @param {string}                [props.role]          The ARIA role for the menu item.
  *
  * @example
- * <caption>ES5</caption>
```
Comment (gziolo) [inline packages/edit-post/src/components/block-settings-menu/plugin-block-settings-menu-item.js:43]: @nosolosw, wasn't it necessary for the Block Editor Handbook to add tabs to switch between two versions of the example?

Comment (gziolo) [reply packages/edit-post/src/components/block-settings-menu/plugin-block-settings-menu-item.js:43]: Hmm, I don't see it having any impact on the generated output. So it looks like it was obsolete 😄

Comment (nosolosw) [reply packages/edit-post/src/components/block-settings-menu/plugin-block-settings-menu-item.js:43]: I commented in another thread above I can't find now :sweat_smile: That was the original goal but don't remember we ever implemented it.

**Comment by sarayourfriend** [comment]
@gziolo @nosolosw Is this PR ready to merge for y'all? I'd like to avoid any more merge conflicts if possible 😁

**Comment by gziolo** [comment]
I like all the fixes, I'm happy to approve 😃

@nosolosw, any blockers?

**Review by gziolo — APPROVED** [review]

File: packages/docgen/test/fixtures/default-parse-error/code.js [hunk 17]
```diff
@@ -1,9 +1,11 @@
+/* eslint-disable jsdoc/valid-types */
 /**
  * Function invoking callback after delay with current timestamp in milliseconds
  * since epoch.
  *
- * @param {(timestamp:number)=>void} callback Callback function.
+ * @param {(timestamp:number=>void} callback Callback function.
```
Comment (nosolosw) [inline packages/docgen/test/fixtures/default-parse-error/code.js:6]:
```suggestion
 * @param {(timestamp:number)=>void} callback Callback function.
```

**Comment by nosolosw** [comment]
Please, do merge! I had already approved a week ago :D

**Pull request merged**